File: ml-fastvlm/app/FastVLM App/Iris/Tools.swift
//
// Tools.swift
// Swift port of tools.py — calculator, complex matrix multiplier, and
// multi-source web search. Each function is `async throws` and returns
// a plain String the UI can render directly.
//

import Foundation

enum IrisToolError: Error, LocalizedError {
    case noMathExpression
    case evalFailure(String)
    case matrixParseFailure(String)
    case networkFailure(String)

    var errorDescription: String? {
        switch self {
        case .noMathExpression:        return "No math expression found."
        case .evalFailure(let s):      return "Eval error: \(s)"
        case .matrixParseFailure(let s): return "Matrix parse error: \(s)"
        case .networkFailure(let s):   return "Network error: \(s)"
        }
    }
}

enum Tools {

    // ─────────────────────────────────────────────────────────────
    // Tool 1: Scalar calculator (NSExpression — sandboxed, no `eval`).
    // ─────────────────────────────────────────────────────────────

    /// Evaluate `45 * 12`, `100 / 4 + 3`, etc. Strips everything outside
    /// `[0-9 + - * / . ( )]` first to keep NSExpression away from any
    /// keypath-style hostile input.
    static func calculator(_ raw: String) -> String {
        let safeChars = CharacterSet(charactersIn: "0123456789+-*/.() ")
        let safe = raw.unicodeScalars.filter { safeChars.contains($0) }
        let expr = String(String.UnicodeScalarView(safe))
            .trimmingCharacters(in: .whitespaces)

        guard !expr.isEmpty else {
            return "Error: No math expression found in '\(raw)'"
        }

        let nsExpr = NSExpression(format: expr)
        guard let result = nsExpr.expressionValue(with: nil, context: nil) else {
            return "Error evaluating '\(expr)'"
        }
        // NSExpression returns NSNumber for numerics. Format ints cleanly.
        if let n = result as? NSNumber {
            let d = n.doubleValue
            if d == d.rounded() && abs(d) < 1e15 {
                return String(Int64(d))
            }
            return String(d)
        }
        return "\(result)"
    }

    // ─────────────────────────────────────────────────────────────
    // Tool 2: Complex matrix multiplier.
    // ─────────────────────────────────────────────────────────────

    /// Parse all numbers (real + complex) out of `problemText`, split into
    /// two equal-size square matrices, multiply, and pretty-print.
    /// Mirrors execute_matrix_multiply in tools.py.
    static func matrixMultiply(_ problemText: String) -> String {
        // Numbers: optional sign, digits, optional decimal, optional `i`.
        // Pulls "1+i", "4-i", "2", "i" etc.
        let numberPattern = #"(?<![A-Za-z])(?:[+\-]?\s*\d+(?:\.\d+)?\s*[+\-]\s*\d*\s*i|[+\-]?\s*\d+(?:\.\d+)?|[+\-]?\s*\d*\s*i)(?![A-Za-z])"#

        guard let regex = try? NSRegularExpression(pattern: numberPattern) else {
            return "Error: matrix regex failed."
        }
        let nsRange = NSRange(problemText.startIndex..<problemText.endIndex, in: problemText)
        let matches = regex.matches(in: problemText, range: nsRange)

        let raw: [String] = matches.compactMap {
            guard let r = Range($0.range, in: problemText) else { return nil }
            let s = String(problemText[r]).trimmingCharacters(in: .whitespaces)
            return s.isEmpty || s == "+" || s == "-" ? nil : s
        }
        let parsed: [Complex] = raw.map { parseComplex($0) }

        guard parsed.count >= 4 else {
            return "Error: Could not extract enough numbers (got \(parsed.count))."
        }

        // Half goes to A, half to B.
        let half = parsed.count / 2
        let aNums = Array(parsed.prefix(half))
        let bNums = Array(parsed.suffix(parsed.count - half))

        let size = Int(Double(aNums.count).squareRoot().rounded())
        guard size * size == aNums.count, size * size == bNums.count else {
            return "Error: Matrix is not square (found \(aNums.count) elements per matrix)."
        }

        var c = [[Complex]](repeating: [Complex](repeating: .zero, count: size), count: size)
        for i in 0..<size {
            for j in 0..<size {
                var s = Complex.zero
                for k in 0..<size {
                    s = s + aNums[i * size + k] * bNums[k * size + j]
                }
                c[i][j] = s
            }
        }

        var lines: [String] = ["C ="]
        for row in c {
            let cells = row.map { format($0) }.joined(separator: ",  ")
            lines.append("  [\(cells)]")
        }
        return lines.joined(separator: "\n")
    }

    // ─────────────────────────────────────────────────────────────
    // Tool 3: Web search (Wikipedia → Tavily fallback).
    // ─────────────────────────────────────────────────────────────

    /// Multi-source factual lookup. Wikipedia first (free, fast,
    /// definition-quality). Tavily second (LLM-curated snippets, requires
    /// `TavilyAPIKey` in Info.plist).
    ///
    /// DuckDuckGo (the third source in tools.py) is intentionally omitted:
    /// its iOS-friendly REST surface is unstable and Tavily is strictly
    /// better when the key is present.
    static func webSearch(_ query: String) async -> String {
        let cleaned = cleanQuery(query)
        guard !cleaned.isEmpty else { return "Error: empty search query." }

        let isDefinition = ["what is", "who is", "define", "meaning of"]
            .contains { query.lowercased().contains($0) }

        // 1. Wikipedia.
        if isDefinition,
           let wiki = await wikipediaSummary(for: cleaned) {
            return "[Wikipedia] \(wiki)"
        }

        // 2. Tavily.
        if let key = IrisConfig.tavilyAPIKey,
           let tav = await tavilySearch(query: query, key: key) {
            return "[Tavily] \(tav)"
        }

        // Last-resort Wikipedia even for non-definition queries.
        if let wiki = await wikipediaSummary(for: cleaned) {
            return "[Wikipedia] \(wiki)"
        }

        return "Sorry, I couldn't find a specific answer for '\(cleaned)'."
    }

    // MARK: - Web helpers

    private static func cleanQuery(_ raw: String) -> String {
        // Strip embedded JSON noise.
        var s = raw
        if let range = s.range(of: #"\{.*?\}"#, options: .regularExpression) {
            s.removeSubrange(range)
        }
        // Strip leading interrogatives.
        let prefixes = ["what is ", "who is ", "where is ", "when was ",
                        "tell me about ", "search for ", "how much is ",
                        "current price of ", "define "]
        let lower = s.lowercased()
        for p in prefixes where lower.hasPrefix(p) {
            s = String(s.dropFirst(p.count))
            break
        }
        return s.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "?"))
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func wikipediaSummary(for query: String) async -> String? {
        let slug = query.replacingOccurrences(of: " ", with: "_")
        guard let encoded = slug.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed),
              let url = URL(string: "https://en.wikipedia.org/api/rest_v1/page/summary/\(encoded)")
        else { return nil }

        var req = URLRequest(url: url, timeoutInterval: 5)
        req.setValue("Iris/1.0 (iOS)", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await URLSession.shared.data(for: req)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200,
                  let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let extract = json["extract"] as? String,
                  !extract.isEmpty
            else { return nil }
            return extract
        } catch {
            return nil
        }
    }

    private static func tavilySearch(query: String, key: String) async -> String? {
        guard let url = URL(string: "https://api.tavily.com/search") else { return nil }
        var req = URLRequest(url: url, timeoutInterval: 8)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "api_key": key,
            "query": query,
            "search_depth": "basic",
            "max_results": 1
        ]
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)

        do {
            let (data, response) = try await URLSession.shared.data(for: req)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200,
                  let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let results = json["results"] as? [[String: Any]],
                  let first = results.first,
                  let content = first["content"] as? String,
                  !content.isEmpty
            else { return nil }
            return content
        } catch {
            return nil
        }
    }

    // MARK: - Complex math (no swift-numerics dep)

    struct Complex: Equatable {
        var real: Double
        var imag: Double

        static let zero = Complex(real: 0, imag: 0)

        static func + (a: Complex, b: Complex) -> Complex {
            Complex(real: a.real + b.real, imag: a.imag + b.imag)
        }
        static func * (a: Complex, b: Complex) -> Complex {
            Complex(
                real: a.real * b.real - a.imag * b.imag,
                imag: a.real * b.imag + a.imag * b.real
            )
        }
    }

    private static func format(_ c: Complex) -> String {
        let r = roundClean(c.real)
        let i = roundClean(c.imag)
        if i == 0 { return formatNumber(r) }
        let sign = i >= 0 ? "+" : "-"
        return "\(formatNumber(r)) \(sign) \(formatNumber(abs(i)))i"
    }

    private static func roundClean(_ v: Double) -> Double {
        let r = (v * 10000).rounded() / 10000
        return r == 0 ? 0 : r
    }

    private static func formatNumber(_ v: Double) -> String {
        if v == v.rounded() { return String(Int64(v)) }
        return String(v)
    }

    /// Parse `1+i`, `4-i`, `2`, `i`, `3.5`, `-2i`, etc.
    private static func parseComplex(_ raw: String) -> Complex {
        var s = raw.replacingOccurrences(of: " ", with: "")
        if s.isEmpty { return .zero }

        // Plain number, no `i` at all.
        if !s.contains("i") {
            return Complex(real: Double(s) ?? 0, imag: 0)
        }

        // Pure imaginary (no real part).
        // Match "i", "+i", "-i", "2i", "-3i", "1.5i".
        if !s.contains("+") || s.firstIndex(of: "+") == s.startIndex {
            // No real part: try to extract leading sign + magnitude.
            return Complex(real: 0, imag: parseImagMagnitude(s))
        }
        if s.dropFirst().contains("-") == false && s.contains("i") && !s.contains("+") {
            return Complex(real: 0, imag: parseImagMagnitude(s))
        }

        // Real ± imag form, e.g. "1+i", "4-2i", "-3.5+i".
        // Find the operator that separates the two parts. Skip a leading
        // sign on position 0.
        var splitIdx: String.Index?
        var idx = s.index(after: s.startIndex)
        while idx < s.endIndex {
            let ch = s[idx]
            if ch == "+" || ch == "-" {
                splitIdx = idx
                break
            }
            idx = s.index(after: idx)
        }
        guard let cut = splitIdx else {
            return Complex(real: 0, imag: parseImagMagnitude(s))
        }

        let realStr = String(s[s.startIndex..<cut])
        let imagStr = String(s[cut..<s.endIndex])
        return Complex(
            real: Double(realStr) ?? 0,
            imag: parseImagMagnitude(imagStr)
        )
    }

    /// "i" -> 1, "-i" -> -1, "2i" -> 2, "-3.5i" -> -3.5.
    private static func parseImagMagnitude(_ raw: String) -> Double {
        var s = raw
        guard let iRange = s.range(of: "i") else { return 0 }
        s.removeSubrange(iRange)
        if s.isEmpty || s == "+" { return 1 }
        if s == "-" { return -1 }
        return Double(s) ?? 0
    }
}

File: ml-fastvlm/tools.py
import re
import numpy as np

# ─────────────────────────────────────────────
# Tool 1: Scalar Calculator
# ─────────────────────────────────────────────
def execute_calculator(equation: str) -> str:
    """Safe scalar calculator. e.g. "45 * 12", "100 / 4 + 3" """
    math_match = re.search(r'[\d\s\+\-\*\/\.\(\)]+', equation)
    if not math_match:
        return f"Error: No math expression found in '{equation}'"
    safe_equation = math_match.group(0).strip()
    if not safe_equation:
        return f"Error: Empty expression from '{equation}'"
    try:
        import ast, operator
        _op_map = {ast.Add: operator.add, ast.Sub: operator.sub, ast.Mult: operator.mul, ast.Div: operator.truediv, ast.USub: operator.neg, ast.UAdd: operator.pos}
        def _eval(node):
            if isinstance(node, ast.Expression): return _eval(node.body)
            elif isinstance(node, ast.Constant): return node.value
            elif isinstance(node, ast.BinOp): return _op_map[type(node.op)](_eval(node.left), _eval(node.right))
            elif isinstance(node, ast.UnaryOp): return _op_map[type(node.op)](_eval(node.operand))
            else: raise ValueError("Unsupported syntax")
        result = _eval(ast.parse(safe_equation, mode='eval'))
        return str(result)
    except Exception as e:
        return f"Error evaluating '{safe_equation}': {e}"


# ─────────────────────────────────────────────
# Tool 2: NumPy Matrix Multiplier
# ─────────────────────────────────────────────
def _parse_complex(s: str) -> complex:
    """Parse complex number strings like 1+i, 4-i, 2, i, 3."""
    s = s.strip().replace(' ', '')
    s = re.sub(r'(?<![0-9])i', '1j', s)
    s = re.sub(r'([0-9])i', r'\1j', s)
    try:
        return complex(s)
    except Exception:
        return complex(0)


def execute_matrix_multiply(problem_text: str) -> str:
    """
    Parses a matrix multiplication problem and computes C = A @ B using NumPy.
    Supports complex numbers using 'i' notation (1+i, 4-i, etc).
    """
    try:
        all_nums = re.findall(
            r'(?<![a-zA-Z])(?:[+\-]?\s*\d+(?:\.\d+)?\s*[+\-]\s*\d*\s*i|[+\-]?\s*\d+(?:\.\d+)?|[+\-]?\s*\d*\s*i)(?![a-zA-Z])',
            problem_text
        )
        filtered = [n for n in all_nums if n.strip() and n.strip() not in ('+', '-')]
        parsed = [_parse_complex(n) for n in filtered]
        if len(parsed) < 4:
            return f"Error: Could not extract enough numbers. Found: {parsed}"
        half = len(parsed) // 2
        a_nums = parsed[:half]
        b_nums = parsed[half:]
        size = int(len(a_nums) ** 0.5)
        if size * size != len(a_nums):
            return f"Error: Matrix is not square (found {len(a_nums)} elements per matrix)"
        A = np.array(a_nums, dtype=complex).reshape(size, size)
        B = np.array(b_nums, dtype=complex).reshape(size, size)
        C = A @ B
        def fmt(v):
            v = complex(v)
            r = round(v.real, 4)
            im = round(v.imag, 4)
            if im == 0:
                return str(int(r) if r == int(r) else r)
            sign = '+' if im >= 0 else '-'
            return f"{int(r) if r == int(r) else r} {sign} {abs(int(im) if im == int(im) else im)}i"
        rows = []
        for row in C:
            rows.append('[' + ',  '.join(fmt(v) for v in row) + ']')
        return 'C =\n  ' + '\n  '.join(rows)
    except Exception as e:
        return f"Error computing matrix multiplication: {e}"


# ─────────────────────────────────────────────
# Tool 3: Web Search (Powered by Tavily AI)
# ─────────────────────────────────────────────
def execute_web_search(query: str) -> str:
    """
    Perform a high-precision web search using Tavily AI.
    Replaces Wikipedia and DuckDuckGo for better accuracy.
    """
    import os
    import requests
    from dotenv import load_dotenv
    load_dotenv()

    api_key = os.getenv("TAVILY_API_KEY")
    if not api_key:
        return "Error: TAVILY_API_KEY not found in environment."

    # Clean the query (remove any JSON artifacts if the LLM leaked them)
    if '{' in query:
        # Try to extract content between quotes if it looks like JSON
        match = re.search(r'"(?:text_in_image|query)":\s*"([^"]+)"', query)
        if match:
            query = match.group(1)
        else:
            # Fallback: just remove curly braces and extra quotes
            query = query.replace('{', '').replace('}', '').replace('"', '').strip()
    
    query = query.strip()
    if not query:
        return "Error: Cleaned search query is empty."

    print(f"DEBUG: Sending to Tavily -> '{query}'")
    try:
        # Minimal payload for maximum compatibility across all API key tiers
        url = "https://api.tavily.com/search"
        payload = {
            "api_key": api_key,
            "query": query,
            "max_results": 1
        }
        response = requests.post(url, json=payload, timeout=15)
        
        if response.status_code == 200:
            data = response.json()
            results = data.get("results", [])
            if results:
                content = results[0].get('content', '')
                return f"\U0001f310 Tavily: {content}"
        
        # If Tavily fails or returns nothing, fallback to DuckDuckGo
        return execute_fallback_search(query)
            
    except Exception as e:
        print(f"Tavily Error: {e}")
        return execute_fallback_search(query)

def execute_fallback_search(query: str) -> str:
    """Zero-config fallback using DuckDuckGo."""
    try:
        import requests
        from bs4 import BeautifulSoup
        headers = {'User-Agent': 'Mozilla/5.0'}
        url = f"https://html.duckduckgo.com/html/?q={query}"
        resp = requests.get(url, headers=headers, timeout=5)
        if resp.status_code == 200:
            soup = BeautifulSoup(resp.text, 'html.parser')
            result = soup.find('a', class_='result__a')
            snippet = soup.find('a', class_='result__snippet')
            if result and snippet:
                return f"\U0001f50e Web (Fallback): {snippet.get_text()}"
    except Exception:
        pass
    return f"Sorry, I couldn't find a specific answer for '{query}'."


# ─────────────────────────────────────────────
# Tool Registry — Add new tools here
# ─────────────────────────────────────────────
AVAILABLE_TOOLS = {
    "calculator": execute_calculator,
    "matrix":     execute_matrix_multiply,
    "web_search": execute_web_search,
}